firmware/library/L1_Drivers/pwm.hpp

```diff
@@ -11,3 +14,129 @@ class PwmInterface
+        static_assert(pwm_pin <= 5,
+                      "Available PWM pins only go up to pin 5.");
+        return Pwm(pwm_port, pwm_pin);
+    }
+
+    Pwm(uint8_t vpport, uint8_t vppin):
+        pwm_(pwm_pin_),
+        pwm_pin_(vpport, vppin)
+    {
+        // sets port to proper function and get pwm channel
+        if (pwm_pin_.GetPort() == 2)
+        {
+            // 0b010 is PWM
+            pwm_pin_.SetPinFunction(kPwmFunction);
+            port = 1;
+            channel = static_cast<uint8_t>(pwm_pin_.GetPin() + 1);
+        }
+        duty_cycle = 0;
+    }
+
+    ~Pwm()
+    {
+        // 0b000 is GPIO
+        pwm_pin_.SetPinFunction(kDefaultFunction);
+    }
+
+    void Initialize(uint32_t frequency_hz = 1'000) override
+    {
+        // DIRECT MANIUPULATION OF SYSTEM CLOCK REGISTERS
+        // WILL REPLACE WITH FUNCTIONS LATER
+        // Enables PWM1 power/clock control bit
+        LPC_SC->PCONP |= (1 << 6);
+        // input clock is divided by 1
+        LPC_SC->PCLKSEL = (1 << 0);
+        // END OF DIRECT MANIPULATION
+
+        match_value = kDefaultClockSpeed / frequency_hz;
+
+        // Resets PWMTC on Match with MR0
+        LPC_PWM1->MCR |= (1 << 1);
+        LPC_PWM1->MR0 = match_value;
+
+        // Enables PWM TC and PC for counting and enables PWM mode
+        LPC_PWM1->TCR = (1<< 0) | (1 << 3);
+        LPC_PWM1->CTCR &= ~(0xF << 0);
+
+
+        // Enables PWM[channel] output
+        LPC_PWM1->LER |= ~(1 << 0);
+        LPC_PWM1->PCR |= (1 << (channel + 8));
+    }
+
+    void SetDutyCycle(float percentage) override
+    {
+        if (percentage < 0 || percentage > 100)
+        {
+            return;
+        }
+        duty_cycle = static_cast<uint32_t>((percentage * static_cast<float>(match_value)) / 100);
+
+        switch (channel)
+        {
+            case 1: LPC_PWM1->MR1 = duty_cycle;
+            case 2: LPC_PWM1->MR2 = duty_cycle;
+            case 3: LPC_PWM1->MR3 = duty_cycle;
+            case 4: LPC_PWM1->MR4 = duty_cycle;
+            case 5: LPC_PWM1->MR5 = duty_cycle;
+            case 6: LPC_PWM1->MR6 = duty_cycle;
+            default: return;
+        }
+        LPC_PWM1->LER |= (1 << channel);
+    }
+
+    float GetDutyCycle() override
+    {
+        return static_cast<float>(duty_cycle / match_value);
+    }
+
+    void SetFrequency(uint32_t frequency_hz) override
+    {
+        // disables PWM mode; this will reset all counters to 0
+        // and allow us to update MR0
+        LPC_PWM1->TCR &= ~(1 << 3);
+
+        match_value = kDefaultClockSpeed / frequency_hz;
+        LPC_PWM1->MR0 = match_value;
+
+        // re-enable PWM mode
+        LPC_PWM1->LER |= ~(1 << 0);
+        LPC_PWM1->TCR |= (1 << 3);
+    }
+
+    uint32_t GetFrequency() override
+    {
+        return kDefaultClockSpeed / match_value;
+    }
+
+ private:
+    PinConfigureInterface & pwm_;
+
+    PinConfigure pwm_pin_;
+
+    static uint32_t match_value;
```

Comment:
You should make an getter for this match value just in case you need it for testing.